File: structural/composite/file_system/go/file_system/file_system_test.go
package file_system

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestFile_Properties(t *testing.T) {
	file := NewFile("test.txt", 100)
	assert.Equal(t, "test.txt", file.GetName())
	assert.Equal(t, 100, file.GetSize())
}

func TestDirectory_Properties(t *testing.T) {
	dir := NewDirectory("docs")
	assert.Equal(t, "docs", dir.GetName())
	// Initial size should be 0
	assert.Equal(t, 0, dir.GetSize())
	// Initial children slice should be non-nil and empty
	assert.NotNil(t, dir.children)
	assert.Empty(t, dir.children)
}

func TestDirectory_AddAndGetChild(t *testing.T) {
	dir := NewDirectory("docs")
	file1 := NewFile("file1.txt", 50)
	file2 := NewFile("file2.txt", 150)

	dir.Add(file1)
	dir.Add(file2)

	assert.Equal(t, file1, dir.GetChild(0))
	assert.Equal(t, file2, dir.GetChild(1))
	assert.Nil(t, dir.GetChild(2), "Accessing out of bounds index should return nil")
	assert.Nil(t, dir.GetChild(-1), "Accessing negative index should return nil")
	assert.Len(t, dir.children, 2)
}

func TestDirectory_RemoveChild(t *testing.T) {
	dir := NewDirectory("docs")
	file1 := NewFile("file1.txt", 50)
	file2 := NewFile("file2.txt", 150)
	file3 := NewFile("other.txt", 200) // A file not added initially

	dir.Add(file1)
	dir.Add(file2)

	assert.Len(t, dir.children, 2)
	assert.Equal(t, file1, dir.GetChild(0))
	assert.Equal(t, file2, dir.GetChild(1))

	// Remove the first element
	removed := dir.Remove(file1)
	assert.True(t, removed, "Remove should return true when element is found")
	assert.Len(t, dir.children, 1)
	assert.Equal(t, file2, dir.GetChild(0), "file2 should now be at index 0")
	assert.Nil(t, dir.GetChild(1))

	// Try removing again (should fail)
	removed = dir.Remove(file1)
	assert.False(t, removed, "Remove should return false for already removed element")
	assert.Len(t, dir.children, 1)

	// Try removing an element never added
	removed = dir.Remove(file3)
	assert.False(t, removed, "Remove should return false for element never added")
	assert.Len(t, dir.children, 1)

	// Remove the remaining element
	removed = dir.Remove(file2)
	assert.True(t, removed)
	assert.Empty(t, dir.children)
	assert.Nil(t, dir.GetChild(0))
}

func TestDirectory_CalculateSize_Simple(t *testing.T) {
	dir := NewDirectory("docs")
	file1 := NewFile("file1.txt", 50)
	file2 := NewFile("file2.txt", 150)

	dir.Add(file1)
	dir.Add(file2)

	assert.Equal(t, 50+150, dir.GetSize())
}

func TestDirectory_CalculateSize_Nested(t *testing.T) {
	root := NewDirectory("root")
	docs := NewDirectory("docs")
	pics := NewDirectory("pics")
	private := NewDirectory("private")

	fileR1 := NewFile("root_file.log", 10)
	fileD1 := NewFile("doc1.txt", 100)
	fileD2 := NewFile("doc2.pdf", 200)
	fileP1 := NewFile("pic1.jpg", 500)
	filePr1 := NewFile("secret.dat", 1000)

	root.Add(fileR1)
	root.Add(docs)
	root.Add(pics)

	docs.Add(fileD1)
	docs.Add(fileD2)
	docs.Add(private)

	pics.Add(fileP1)

	private.Add(filePr1)

	// Check individual sizes
	assert.Equal(t, 10, fileR1.GetSize())
	assert.Equal(t, 100, fileD1.GetSize())
	assert.Equal(t, 200, fileD2.GetSize())
	assert.Equal(t, 500, fileP1.GetSize())
	assert.Equal(t, 1000, filePr1.GetSize())

	// Check directory sizes
	assert.Equal(t, 1000, private.GetSize())
	assert.Equal(t, 500, pics.GetSize())
	assert.Equal(t, 100+200+1000, docs.GetSize())
	assert.Equal(t, 10+(100+200+1000)+500, root.GetSize())
}

// Display method testing is often limited to ensuring no panics,
// unless output capturing is implemented.
func TestDisplay_RunsWithoutPanic(t *testing.T) {
	root := NewDirectory("root")
	docs := NewDirectory("docs")
	file1 := NewFile("file1.txt", 50)
	root.Add(docs)
	docs.Add(file1)

	assert.NotPanics(t, func() {
		root.Display("")
	})
}

File: structural/composite/file_system/typescript/file_system.test.ts
import { File, Directory, FileSystemComponent } from "./file_system";

describe("FileSystem Composite Tests", () => {
  describe("File (Leaf)", () => {
    let file: File;

    beforeEach(() => {
      file = new File("test.txt", 100);
    });

    test("should have correct name and size", () => {
      expect(file.getName()).toBe("test.txt");
      expect(file.getSize()).toBe(100);
    });

    test("should throw error when trying to add child", () => {
      const anotherFile = new File("another.txt", 50);
      expect(() => file.add(anotherFile)).toThrow(
        "Cannot add to this component"
      );
    });

    test("should throw error when trying to remove child", () => {
      const anotherFile = new File("another.txt", 50);
      // Need to add first to attempt removal, which itself will fail
      // So we test the method directly
      expect(() => file.remove(anotherFile)).toThrow(
        "Cannot remove from this component"
      );
    });

    test("should throw error when trying to get child", () => {
      expect(() => file.getChild(0)).toThrow(
        "Cannot get child from this component"
      );
    });
  });

  describe("Directory (Composite)", () => {
    let directory: Directory;
    let file1: File;
    let file2: File;

    beforeEach(() => {
      directory = new Directory("docs");
      file1 = new File("file1.txt", 50);
      file2 = new File("file2.txt", 150);
    });

    test("should have correct name and initial size 0", () => {
      expect(directory.getName()).toBe("docs");
      expect(directory.getSize()).toBe(0);
    });

    test("should add and get children correctly", () => {
      directory.add(file1);
      directory.add(file2);
      expect(directory.getChild(0)).toBe(file1);
      expect(directory.getChild(1)).toBe(file2);
      expect(() => directory.getChild(2)).toThrow(RangeError); // Check bounds
    });

    test("should remove children correctly", () => {
      directory.add(file1);
      directory.add(file2);
      expect(directory.getChild(0)).toBe(file1);
      expect(directory.getChild(1)).toBe(file2);

      directory.remove(file1); // Remove first child
      expect(directory.getChild(0)).toBe(file2); // Second child is now first
      expect(() => directory.getChild(1)).toThrow(RangeError);

      // Test removing non-existent child
      expect(() => directory.remove(file1)).toThrow(
        "Component 'file1.txt' not found in directory 'docs'"
      );
    });

    test("should calculate size correctly with only files", () => {
      directory.add(file1);
      directory.add(file2);
      expect(directory.getSize()).toBe(50 + 150);
    });

    test("should calculate size correctly with nested directories", () => {
      const root = new Directory("root");
      const docs = new Directory("docs");
      const pics = new Directory("pics");
      const privateDir = new Directory("private");

      const file_r1 = new File("root_file.log", 10);
      const file_d1 = new File("doc1.txt", 100);
      const file_d2 = new File("doc2.pdf", 200);
      const file_p1 = new File("pic1.jpg", 500);
      const file_pr1 = new File("secret.dat", 1000);

      root.add(file_r1);
      root.add(docs);
      root.add(pics);

      docs.add(file_d1);
      docs.add(file_d2);
      docs.add(privateDir);

      pics.add(file_p1);

      privateDir.add(file_pr1);

      // Check individual sizes
      expect(file_r1.getSize()).toBe(10);
      expect(file_d1.getSize()).toBe(100);
      expect(file_d2.getSize()).toBe(200);
      expect(file_p1.getSize()).toBe(500);
      expect(file_pr1.getSize()).toBe(1000);

      // Check directory sizes
      expect(privateDir.getSize()).toBe(1000);
      expect(pics.getSize()).toBe(500);
      expect(docs.getSize()).toBe(100 + 200 + 1000);
      expect(root.getSize()).toBe(10 + (100 + 200 + 1000) + 500);
    });

    test("display method should run without errors", () => {
      // Mock console.log to prevent output during tests (optional)
      const consoleSpy = jest
        .spyOn(console, "log")
        .mockImplementation(() => {});

      const root = new Directory("root");
      const docs = new Directory("docs");
      const file1 = new File("file1.txt", 50);
      root.add(docs);
      docs.add(file1);

      expect(() => root.display()).not.toThrow();

      // Restore console.log
      consoleSpy.mockRestore();
    });
  });
});
